// sass-module-importer/lib/index.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var fs = _interopDefault(require('fs'));
var path = _interopDefault(require('path'));
var Map = _interopDefault(require('es6-map'));
var assign = _interopDefault(require('object-assign'));
var npmResolve = _interopDefault(require('resolve'));
var bowerResolve = _interopDefault(require('resolve-bower'));

var ModuleImporter = function ModuleImporter(opts) {
  this.aliases = new Map();
  this.options = assign({}, { packageFilter: this.filter }, opts);
};

ModuleImporter.prototype.resolve = function resolve(ref) {
  var this$1 = this;
    var url = ref.url;
    var prev = ref.prev;

    if (this.aliases.has(url)) {
    return Promise.resolve(this.aliases.get(url));
  }

  return Promise.resolve({ url: url, prev: prev })
    .then(function ( file ) { return this$1.npm(file); })
    .then(function ( file ) { return this$1.bower(file); })
    .then(function ( file ) { return this$1.read(file); })
    .then(function (res) {
      this$1.aliases.set(url, res);
      return res;
    });
};

ModuleImporter.prototype.filter = function filter(pkg) {
  if (!pkg.main || (pkg.main && !pkg.main.match(/\.s?[c|a]ss$/g))) {
    pkg.main = pkg.style || pkg['main.scss'] || pkg['main.sass'] || 'index.css';
  }
  return pkg;
};

ModuleImporter.prototype.find = function find(resolver, ref) {
  var this$1 = this;
    var url = ref.url;
    var prev = ref.prev;
    var resolved = ref.resolved;

    return new Promise(function (resolve) {
    if (resolved) {
      resolve({ url: url, prev: prev, resolved: resolved });
    } else {
      resolver(url, this$1.options, function (err, res) {
        resolve({ url: (err ? url : res), prev: prev, resolved: !err });
      });
    }
  });
};

ModuleImporter.prototype.read = function read(ref) {
  var url = ref.url;
    var prev = ref.prev;
    var resolved = ref.resolved;

    return new Promise(function (resolve, reject) {
    if (url.match(/\.css$/g)) {
      fs.readFile(url, 'utf8', function (err, contents) {
        if (err) {
          reject(err);
        } else {
          resolve({ contents: contents });
        }
      });
    } else {
      var resolvedURL = url;
      if (!resolved && prev && prev !== 'stdin' && !path.isAbsolute(url)) {
        resolvedURL = path.resolve(path.dirname(prev), url);
      }
      resolve({ file: resolvedURL });
    }
  });
};

ModuleImporter.prototype.npm = function npm(file) {
  return this.find(npmResolve, file);
};

ModuleImporter.prototype.bower = function bower(file) {
  return this.find(bowerResolve, file);
};


/**
 * Look for Sass files installed through npm
 * @param opts {Object}       Options to be passed to the resolver module
 *
 * @return {Function}         Function to be used by node-sass importer
 */
function index (opts) {
  var importer = new ModuleImporter(opts);

  return function (url, prev, done) {
    importer.resolve({ url: url, prev: prev }).then(done);
  };
}

module.exports = index;
